pcli/src/command/stake.rs

```diff
@@ -0,0 +1,145 @@
+        amount: String,
+        /// The transaction fee (paid in upenumbra).
+        #[structopt(long, default_value = "0")]
+        fee: u64,
+        /// Optional. Only spend funds originally received by the given address index.
+        #[structopt(long)]
+        source: Option<u64>,
+    },
+    /// Withdraw stake from a validator's delegation pool.
+    Undelegate {
+        /// The identity key of the validator to withdraw delegation from.
+        #[structopt(long)]
+        from: String,
+        /// The amount of stake to delegate.
+        amount: String,
+        /// The transaction fee (paid in upenumbra).
+        #[structopt(long, default_value = "0")]
+        fee: u64,
+        /// Optional. Only spend funds originally received by the given address index.
+        #[structopt(long)]
+        source: Option<u64>,
+    },
+    /// Redelegate stake from one validator's delegation pool to another.
+    Redelegate {
+        /// The identity key of the validator to withdraw delegation from.
+        #[structopt(long)]
+        from: String,
+        /// The identity key of the validator to delegate to.
+        #[structopt(long)]
+        to: String,
+        /// The amount of stake to delegate.
+        amount: String,
+        /// The transaction fee (paid in upenumbra).
+        #[structopt(long, default_value = "0")]
+        fee: u64,
+        /// Optional. Only spend funds originally received by the given address index.
+        #[structopt(long)]
+        source: Option<u64>,
+    },
+    /// Display this wallet's delegations and their value.
+    Show,
+    /// Display all of the validators participating in the chain.
+    ListValidators {
+        /// Whether to show validators that are not currently part of the consensus set.
+        #[structopt(long)]
+        show_inactive: bool,
+    },
+}
+
+impl StakeCmd {
+    pub fn needs_sync(&self) -> bool {
+        true
+    }
+
+    pub async fn exec(&self, opt: &Opt, state: &ClientState) -> Result<()> {
+        match self {
+            StakeCmd::Delegate {
+                to,
+                amount,
+                fee,
+                source,
+            } => {
+                let unbonded_amount = {
+                    let Value { amount, asset_id } = amount.parse::<Value>()?;
+                    if asset_id != *STAKING_TOKEN_ASSET_ID {
+                        return Err(anyhow!("staking can only be done with the staking token"));
+                    }
+                    amount
+                };
+
+                let to = to.parse::<IdentityKey>()?;
+
+                // FIXME! need some kind of structure for recording chain
+                // parameters - connected with having protos for genesis data
+                // (though not all genesis data is chain parameters)
+                let epoch_duration = 10;
+                let current_epoch =
+                    Epoch::from_height(state.last_block_height().unwrap() as u64, epoch_duration);
```

Comment:
We don't currently have any kind of structure of recording chain parameters that the client might want to know about. This has already been a problem when trying to use the chain ID to detect wrong-chain errors, since the client state doesn't have anywhere to store the chain ID.

It's connected to the idea of having proto-specified genesis data, since we'd probably want the genesis data to be a superset of the chain parameters (e.g., the chain params include the epoch duration, but not the genesis allocations).

Reply:
This is now possible with #332